src/raftstore/store/fsm/apply.rs

```diff
@@ -2029,196 +2092,479 @@ pub struct ApplyRes {
+impl Fsm for ApplyFsm {
+    type Message = Msg;
+
+    #[inline]
+    fn is_stopped(&self) -> bool {
+        self.delegate.stopped
+    }
+
+    #[inline]
+    fn set_mailbox(&mut self, mailbox: Cow<BasicMailbox<Self>>)
+    where
+        Self: Sized,
+    {
+        self.mailbox = Some(mailbox.into_owned());
+    }
+
+    #[inline]
+    fn take_mailbox(&mut self) -> Option<BasicMailbox<Self>>
+    where
+        Self: Sized,
+    {
+        self.mailbox.take()
+    }
+}
+
+impl Drop for ApplyFsm {
+    fn drop(&mut self) {
+        self.delegate.clear_all_commands_as_stale();
     }
+}
+
+pub struct ControlMsg;
+
+pub struct ControlFsm;
+
+impl Fsm for ControlFsm {
+    type Message = ControlMsg;
+
+    #[inline]
+    fn is_stopped(&self) -> bool {
+        true
+    }
+}
 
-    fn handle_destroy(&mut self, d: Destroy) {
-        // Only respond when the meta exists. Otherwise if destroy is triggered
-        // multiple times, the store may destroy wrong target peer.
-        if let Some(meta) = self.delegates.remove(&d.region_id) {
-            let mut meta = meta.unwrap();
-            info!("{} remove from apply delegates", meta.tag);
-            meta.destroy();
-            self.notifier.send(TaskRes::Destroy(meta)).unwrap();
+pub struct ApplyPoller {
+    msg_buf: Vec<Msg>,
+    apply_ctx: ApplyContext,
+    messages_per_tick: usize,
+}
+
+impl PollHandler<ApplyFsm, ControlFsm> for ApplyPoller {
+    fn begin(&mut self) {}
+
+    /// There is no control fsm in apply poller.
+    fn handle_control(&mut self, _: &mut ControlFsm) -> Option<usize> {
+        unimplemented!()
+    }
+
+    fn handle_normal(&mut self, normal: &mut ApplyFsm) -> Option<usize> {
+        let mut expected_msg_count = None;
+        if normal.delegate.wait_merge_state.is_some() {
+            // We need to query the length first, otherwise there is a race
+            // condition that new messages are queued after resuming and before
+            // query the length.
+            expected_msg_count = Some(normal.receiver.len());
+            if !normal.resume_pending_merge(&mut self.apply_ctx) {
+                return expected_msg_count;
+            }
+            expected_msg_count = None;
+        }
+        while self.msg_buf.len() < self.messages_per_tick {
+            match normal.receiver.try_recv() {
+                Ok(msg) => self.msg_buf.push(msg),
+                Err(TryRecvError::Empty) => {
+                    expected_msg_count = Some(0);
+                    break;
+                }
+                Err(TryRecvError::Disconnected) => {
+                    normal.delegate.stopped = true;
+                    expected_msg_count = Some(0);
+                    break;
+                }
+            }
         }
+        normal.handle_tasks(&mut self.apply_ctx, &mut self.msg_buf);
+        if normal.delegate.merged {
+            normal.delegate.destroy(&mut self.apply_ctx);
+            // Set it to 0 to clear all messages remained in queue.
```

Comment:
Why can `Some(0)` clear remaining messages even after this fsm is stopped?

Reply:
As the trait document says, `Some(0)` means as long as the channel is not empty, this fsm should be kept notified.